crates/task-impls/src/quorum_vote/handlers.rs

```diff
@@ -172,43 +172,67 @@ pub(crate) async fn update_shared_state<
     storage: Arc<RwLock<I::Storage>>,
     proposed_leaf: &Leaf<TYPES>,
     vid_share: &Proposal<TYPES, VidDisperseShare<TYPES>>,
+    parent_view_number: Option<TYPES::View>,
 ) -> Result<()> {
     let justify_qc = &proposed_leaf.justify_qc();
 
+    let consensus_reader = consensus.read().await;
+    // Try to find the validated vview within the validasted state map. This will be present
+    // if we have the saved leaf, but if not we'll get it when we fetch_proposal.
+    let mut maybe_validated_view = match parent_view_number {
+        Some(parent_view_number) => consensus_reader
+            .validated_state_map()
+            .get(&parent_view_number)
+            .cloned(),
+        None => None,
+    };
+
     // Justify qc's leaf commitment should be the same as the parent's leaf commitment.
-    let mut maybe_parent = consensus
-        .read()
-        .await
+    let mut maybe_parent = consensus_reader
         .saved_leaves()
         .get(&justify_qc.data.leaf_commit)
         .cloned();
+
+    drop(consensus_reader);
+
     maybe_parent = match maybe_parent {
         Some(p) => Some(p),
-        None => fetch_proposal(
-            justify_qc.view_number(),
-            sender.clone(),
-            receiver.activate_cloned(),
-            Arc::clone(&quorum_membership),
-            OuterConsensus::new(Arc::clone(&consensus.inner_consensus)),
-            public_key.clone(),
-            private_key.clone(),
-            &upgrade_lock,
-        )
-        .await
-        .ok(),
+        None => {
+            match fetch_proposal(
+                justify_qc.view_number(),
+                sender.clone(),
+                receiver.activate_cloned(),
+                Arc::clone(&quorum_membership),
+                OuterConsensus::new(Arc::clone(&consensus.inner_consensus)),
+                public_key.clone(),
+                private_key.clone(),
+                &upgrade_lock,
+            )
+            .await
+            .ok()
+            {
+                Some((leaf, view)) => {
+                    maybe_validated_view = Some(view);
+                    Some(leaf)
+                }
+                None => None,
+            }
+        }
     };
+
     let parent = maybe_parent.context(info!(
         "Proposal's parent missing from storage with commitment: {:?}, proposal view {:?}",
         justify_qc.data.leaf_commit,
         proposed_leaf.view_number(),
     ))?;
-    let consensus_reader = consensus.read().await;
 
-    let (Some(parent_state), _) = consensus_reader.state_and_delta(parent.view_number()) else {
-        bail!("Parent state not found! Consensus internally inconsistent");
+    let Some(validated_view) = maybe_validated_view else {
+        bail!("Don't have a validated view for our view! Consensus internally inconsistent");
```

Comment:
I think this error is too aggressive, it's possible we just didn't see that view's proposal and failed to fetch it.  Consensus isn't inconsistent in that case.  The error just below is much more serious because it means we stored a view, but not the state update for that view which should never happen